### PR TITLE
[2/x]: fix numerics integration test and test delayed vs dynamic

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #300
* #299
* #298
* #297
* #296
* #294
* #293
* __->__ #291
* #290

Summary:

1. the SAM test wasn't easy to use because it had real weights and hence
   required real data for useful testing, which is not convenient from
   an integration test. Switched to LLaMa FFN with random weights, and
   made all the thresholds tight to actually check numerics are close.
2. extended numerics test to check all combinations of delayed vs
   dynamic
3. to be able to do (2), extended the module swap utility to configure
   delayed vs dynamic on a model level, for now without an option to
   customize further

Test Plan:

```
pytest test/test_numerics_integration.py -s -x
./test/test_everything.sh
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D59305796](https://our.internmc.facebook.com/intern/diff/D59305796)

### CODE REVIEW

File: test/test_everything.sh [hunk 1]
```diff
@@ -5,8 +5,8 @@ set -e
 IS_ROCM=$(rocm-smi --version || true)
 
 pytest test/test_base.py
-pytest test/test_sam.py
```
Comment (drisspg) [inline test/test_everything.sh:8]: love it, this test has been annoying

File: test/test_numerics_integration.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        optim_fp8.zero_grad()
+        if linear_requires_sync(
+            linear_type, scaling_type_x, scaling_type_w, scaling_type_dL_dY
+        ):
+            sync_float8_amax_and_scale_history(model_fp8)
+        model_fp8_out = model_fp8(data2)
+        model_fp8_out.sum().backward()
+
+        out_sqnr = compute_error(model_ref_out, model_fp8_out)
+        assert out_sqnr > 20.0
```
Comment (drisspg) [inline test/test_numerics_integration.py:175]: nit: maybe a message

**Review by drisspg — APPROVED** [review]
great clean up, love it!

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59305796).

**Pull request closed without merging**